### DISCUSSION TITLE
Issues about building own MOOSE application under cardinal environment

### DISCUSSION THREAD

**Discussion by WayneX-Wong** [neams-th-coe/cardinal#529] [Q&A] [answered]
Hi all,

I install cardinal without MOOSE‘s conda environment, here's my PETSc_DIR and LIBMESH_DIR
![image](https://user-images.githubusercontent.com/111510550/205862119-343463c3-4c63-42a3-869f-98b1b6797ca5.png)

I try to build my own MOOSE application under cardinal/contrib environment using method like https://mooseframework.inl.gov/getting_started/new_users.html, 

While when I make, some errors occur like
![image](https://user-images.githubusercontent.com/111510550/205861849-b8cbb520-c244-4ee6-9d13-2683b860314a.png)

I can't see where the problem is, I can actually find libmesh-config.in at /home/lagrange/cardinal/contrib/moose/libmesh/contrib/bin

Is there any possible solution?

Many thanks!

**Comment by aprilnovak** [comment, marked as answer]
Hi @WayneX-Wong, that `libmesh-config` error looks to me like you didn't build libMesh or PETSc. The two simplest options for building a MOOSE application are:

1. Use MOOSE's conda environment
2. Build libMesh and PETSc using scripts. For this option, you would need to have run:

```
./contrib/moose/scripts/update_and_rebuild_petsc.sh
./contrib/moose/scripts/update_and_rebuilld_libmesh.sh
```

Simply setting `LIBMESH_DIR` and `PETSC_DIR` aren't enough to build those dependencies. Which of the two options above would you like to pursue? I can then provide more detailed instructions.

**Reply by WayneX-Wong** [reply]
Thank's for your reply @aprilnovak .  I failed in compiling cardinal using MOOSE's conda environment previously, so I would prefer option 2

While I believe I've ran both scripts when I compiled cardinal without MOOSE's conda environment, and successfully ran some cases using cardinal already. So I'm not sure if it's compilation issue.

Anyway, I'll try option 2 again later. More detailed instructions would be grateful!

**Reply by aprilnovak** [reply]
Hi @WayneX-Wong - it's possible that you did indeed run those scripts successfully, but magpie doesn't know about where you have MOOSE located. Can you try this:

```
export MOOSE_DIR=/path/to/cardinal/contrib/moose
cd cardinal/contrib/magpie
make -j4
```

**Reply by WayneX-Wong** [reply]
Sorry for the late @aprilnovak  , I've set:
`export MOOSE_DIR=/path/to/cardinal/contrib/moose`
`cd cardinal/contrib/magpie`
`make -j4`
But still same errors.

Maybe rebuild Cardinal under MOOSE's conda environment could possibly solve this question?

**Reply by aprilnovak** [reply]
I just want to be super sure we're on the same page. Did you use this verbatim:

```
export MOOSE_DIR=/path/to/cardinal/contrib/moose
```

or did you adjust the path to reflect where you actually have Cardinal. For instance, for my computer, what I meant by `/path/to/cardinal` is `/home/anovak/cardinal`.

**Reply by WayneX-Wong** [reply]
Yeah of course, for me I set:
`export MOOSE_DIR=/home/lagrange/cardinal/contrib/moose`

Sorry for the vague statement I just made

**Reply by aprilnovak** [reply]
Is this the Magpie you're using? https://github.com/idaholab/magpie

I can try to build it myself to see if I can reproduce your error.

**Reply by aprilnovak** [reply]
I'm able to build with that Magpie ^.

I think you'll want to unset LIBMESH_DIR and PETSC_DIR. I think those values aren't being set correctly.

```
unset LIBMESH_DIR
unset PETSC_DIR
cd cardinal/contrib/magpie
make -j8
```

**Reply by WayneX-Wong** [reply]
Ah it just happened to have the same name. I've just ran the script
`./moose/scripts/stork.sh magpie` 
It works, while when I make, the above error occurs.

Whatever it was, with your guidance by unsetting LIBMESH_DIR and PETSC_DIR I successfully build it now!

Thanks again for your patience! @aprilnovak

**Reply by aprilnovak** [reply]
Great to hear, always happy to help!

We've done some recent work trying to make Cardinal easier to build, so it's possible you still had `LIBMESH_DIR` and `PETSC_DIR` being set based on our old instructions. But you shouldn't need those anymore except in a few edge cases.
